### ISSUE TITLE
[Question] Extract one GP from Hadamard Multitask

### ISSUE THREAD

**Issue by irinaespejo** [cornellius-gp/gpytorch#1895]
Hi team :wave: 

# Context
I'm using a Multitask Hadamard GP for Regression as exactly as described in the tutorial [Hadamard Multitask GP Regression](https://docs.gpytorch.ai/en/stable/examples/03_Multitask_Exact_GPs/Hadamard_Multitask_GP_Regression.html) with 2 tasks.

After training, the tutorial proceeds to [evaluate](https://docs.gpytorch.ai/en/stable/examples/03_Multitask_Exact_GPs/Hadamard_Multitask_GP_Regression.html#Make-predictions-with-the-model) both tasks separately with 

```
with torch.no_grad(), gpytorch.settings.fast_pred_var():
      observed_pred_y1 = likelihood(model(test_x, test_i_task1))
      observed_pred_y2 = likelihood(model(test_x, test_i_task2))
```

# Problem
The variables `observed_pred_y1` and `observed_pred_y2` are of the type  `gpytorch.distributions.MultivariateNormal`.
I would like to extract from `model` one GP from one of the tasks, say task 1, called `model1` that is of type `gpytorch.models.ExactGP` instead of a Multivariate distribution. This might sound weird but after training, I need to pass a GP to other custom software and a distribution won't work. It is necessary to use Multitask because both tasks are related and need to be fitted together.

# (failed) Attempts
I've tried to create a class similar to  the classic `ExactGPModel` like [here](https://docs.gpytorch.ai/en/stable/examples/01_Exact_GPs/Simple_GP_Regression.html) but initializing the modules according to the tasK 2 of the Multitask `model`

```
class ExactGPModel_Task1(gpytorch.models.ExactGP):
    def __init__(self, x_train, y_train, likelihood, model):
        super().__init__(x_train, y_train, likelihood)
        self.mean_module = model.mean_module 
        self.covar_module = model.covar_module
        
    def forward(self, x, model):
        mean_x = self.mean_module(x)
        this_test_i_task1 = torch.full((x.shape[0],1), dtype=torch.long, fill_value=0)
        pred_y1 = model(x, this_test_i_task1)
        covar_x = pred_y1._covar
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
 

model1 = ExactGPModel_Task1(train_x1, train_y1, likelihood, model)
```

and when evaluating

```
model1.eval()
likelihood.eval()
model1(test_x, model)
```
I run into  `AttributeError: 'MultitaskGPModel' object has no attribute 'ndimension'` at `gpytorch/module.py in __getattr__(self, name)` starting on this [line](https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/models/exact_gp.py#L244).

Do you have any advice? 

Thank you so much! :)

**Comment by Balandat** [comment]
Could you please post the full stack trace to make it easier to debug?

**Comment by irinaespejo** [comment]
Sure, done. Thanks for the quick reply!

**Comment by wjmaddox** [comment]
I don't think I follow what you're trying to do here. 

Why couldn't you pass in a callable that just explicitly computes the posterior on task 1?

something like:

```python
task1_posterior_fn = lambda test_x: likelihood(model(test_x, test_i_task_1))
```

because having that function would allow you to return the `MultivariateNormal` you're expecting downstream?

**Comment by irinaespejo** [comment]
Hi @wjmaddox, thanks! That would almost work but downstream there are calls like `gp.train_inputs`,  `gp.train_targets`, `gp.parameters` or  `gp.get_fantasy_model`. It is possible to refactor the downstream code so that it works with a Multivariate distribution but I'd rather leave it as last option and try first to accommodate the requirement of input a GP.

**Comment by Balandat** [comment]
@irinaespejo is the downstream code based on BoTorch by any chance?

**Comment by irinaespejo** [comment]
@Balandat Unfortunately not. But we've been considering using BoTorch downstream for a while. How would BoTorch help here?

**Comment by wjmaddox** [comment]
I believe that there's utilities for returning only one task during the posterior call in Botorch's `MultitaskGP` models.

I think I get a sense about what you're trying to do now:

One other thing to try is to define the forwards method of `model1` as taking as input the gp model as a kwarg so this should work (sorry for using botorch model classes but it should be clear how to proceed):

```python
import gpytorch
import torch
from botorch.models import MultiTaskGP

class ExactGPModel_Task1(gpytorch.models.ExactGP):
    def __init__(self, x_train, y_train, likelihood, model):
        super().__init__(x_train, y_train, likelihood)
        self.mean_module = model.mean_module 
        self.covar_module = model.covar_module
        
    def forward(self, x, model):
        mean_x = self.mean_module(x)
        this_test_i_task1 = torch.full((x.shape[0],1), dtype=torch.long, fill_value=0)
        stacked_x = torch.cat((x, this_test_i_task1), dim=-1)
        pred_y1 = model(stacked_x)
        covar_x = pred_y1._covar
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)

train_x1 = torch.rand(10, 3)
train_y1 = torch.rand(10)

train_mtgp = torch.cat((torch.rand(10, 3), torch.randint(3, (10,)).unsqueeze(-1)), dim=-1)

model = MultiTaskGP(train_mtgp, train_y1.unsqueeze(-1), task_feature=-1)
model.eval()

likelihood = gpytorch.likelihoods.GaussianLikelihood()
model1 = ExactGPModel_Task1(train_x1, train_y1, likelihood, model)

model1(train_x1, model=model) # seems to work okay

model1.eval();

test_x1 = torch.rand(10, 3)
model1(test_x1, model=model) # posterior is also okay
```

**Comment by Balandat** [comment]
> I believe that there's utilities for returning only one task during the posterior call in Botorch's MultitaskGP models.

Yeah there is some functionality for a model here: https://github.com/pytorch/botorch/blob/main/botorch/models/multitask.py#L105-L106

I guess I'd be curious if there are some more general patterns in terms of extracting individual tasks that emerge in different setups that we could generalize and provide as utilities as part of botorch.

**Comment by irinaespejo** [comment]
Hi @wjmaddox that worked, thank you! I had to pass the multitask model via kwargs so it wasn't a major refactor.

> I guess I'd be curious if there are some more general patterns in terms of extracting individual tasks that emerge in different setups that we could generalize and provide as utilities as part of botorch.

Personally, I am going to use extracting a task from multitask model quite a lot. I guess one pattern, our pattern, is to use a multitask GP to incorporate correlations of weaker but faster experiments into a high fidelity experiment. And then, use that task to do optimal experiment design. I'd be down to help with this if you decide to support this feature. 

Thanks again!
